### DISCUSSION TITLE
How to add tasmota-2M Version to tasmota menue

### DISCUSSION THREAD

**Discussion by HofeBY** [arendst/Tasmota#22004] [Ideas]
I need the 2M byte version of Tasmota in Nous-A1T adapters and couldn't find the menu item in the compiler menu.

I found the entries in platformio_override.ini
; Build variant 2MB = 1MB firmware, 1MB filesystem (most Shelly devices)
;board = esp8266_2M1M

Then I opened platformio_tasmota_env.ini and added the area between "env:tasmota" & "env:tasmota-4M"

[env:tasmota-2M]
board = esp8266_2M1M
build_flags = ${env.build_flags}

-DCODE_IMAGE_STR='"tasmota-2M"'

-DOTA_URL='"http://ota.tasmota.com/tasmota/release/tasmota-2M.bin.gz"'

The menu item now appears in the compiler menu.

Compilation also runs without any problems.

I think it would make sense to expand this in Tasmota for everyone.

If further changes are necessary, I would be grateful for a tip.

Greetings
HofeBY

**Comment by HofeBY** [comment]
![tasmota-2M-Update](https://github.com/user-attachments/assets/5fcabf81-7ea7-4f7f-86c1-4042132a28be)

**Comment by sfromis** [comment]
The 2M setup can be considered as experimental, and you do **not** need it just because some device may have 2M of flash, where the standard 1M build should work fine. IIRC, you'll even get into trouble if you try to OTA, without also having a minimal to match, or even if you try to upgrade between 1M and 2M builds. I'd not recommend the 2M setup, unless you have a **strong** need for a file system.

**Comment by HofeBY** [comment]
Non functional OTA with tasmota.bin (1M Byte) on this devices is on reason to change to 2M version on possible devices like Nous A1T and most Shelly.
I also need the filesystem to store messurements and larger (8kByte) Script.txt routines.

ESP8266 with external 2M Flash exist since a long time & I think TM is able to support this possibility.

As you see in picture upwards, TM support also 4M Byte external Flash !

And you wrote that it´s experimental, thats also for 4M true, but it´s worth to bring it to fully functional.
Many existing devices will wait for it to be fully functional.
 
So it would be helpfull to integrate 2M in TM and fix if fault´s exist to functional versions.

**Reply by sfromis** [reply]
Now you are boxing yourself into a narrow edge case, instead of arguing for it to be just one of the standard selections. And you've already worked out how to do the build for this setup. But still, it is not unproblematic, unless you are very careful to not mix 1M and 2M setups during upgrading, including needing the minimal to match too.

Obviously, it is fine if you do a build for your preferences, and manage the caveats, but Tasmota is not even trying to provide for the many permutations possible when doing builds.

**Reply by HofeBY** [reply]
I cannot see any edge cases, as many devices with 2M and 4M memory extensions have been around for a long time.

And as you can see, I am open to all suggestions on what else can be improved for the benefit of everyone.

The 4M ESP8266 is already supported by TM, even if it is unusable on my NodeMCU-Amica with 4M flash to constant crashes and restarts.

It runs reasonably normally with 1M, but several functions are missing.

**Reply by sfromis** [reply]
While there indeed are many devices with 4M, that is not a similar situation to the possible trouble specifically for 2M, which is not very common. And most users getting a device which happens to have that odd config (which works "as normal" with standard 1M builds), will not be having your specific preferences for large scripts and local files. Hence "edge case".

Not sure what happens with your NodeMCU and 4M combos, my experience with several NodeMCU boards is that 4M builds are running just fine. And easily swappable to/from 1M builds.

I'm not per se "against" providing for 2M builds without that tiny amount of editing, but the number of use cases is "limited"....

**Comment by s-hadinger** [comment]
There are currently no plans to provide precompiled 2M firmwares; due to combinatorial variety of variants, and the high risk of breaking a working setup if you are not careful enough

**Reply by HofeBY** [reply]
OK, I have to do it by my own and asking for supporters .

**Reply by Jason2866** [reply]
Just a small last note. One mistake providing a false mimimal or firmware variant (swapping 1MB or 2MB build or vice versa) ends in need of serial flashing. This is reason why it is **not** supported.
Tasmota esp8266 OTA with 1MB works perfectly fine.
The esp8266 build variants with 2MB and 4MB do work well too. There are no bugs.
Not supported since there are only very rare seldom use cases where more than 1MB is needed.
The risk of bricking devices is way to high to provide this variants. The build env are in the Platformio setup for compiling if someone has a need for. BUT not everything possible is supported nor recommended ;-)
